src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -159,179 +109,283 @@ public IDisposable On(string methodName, Type[] parameterTypes, Func<object[], o
+        public async Task<ChannelReader<object>> StreamAsChannelAsync(string methodName, Type returnType, object[] args, CancellationToken cancellationToken = default) =>
+            await StreamAsChannelAsyncCore(methodName, returnType, args, cancellationToken).ForceAsync();
+
+        public async Task<object> InvokeAsync(string methodName, Type returnType, object[] args, CancellationToken cancellationToken = default) =>
+            await InvokeAsyncCore(methodName, returnType, args, cancellationToken).ForceAsync();
+
+        // REVIEW: We don't generally use cancellation tokens when writing to a pipe because the asynchrony is only the result of backpressure.
+        // However, this would be the only "invocation" method _without_ a cancellation token... which is odd.
+        public async Task SendAsync(string methodName, object[] args, CancellationToken cancellationToken = default) =>
+            await SendAsyncCore(methodName, args).ForceAsync();
+
+        private async Task StartAsyncCore()
         {
-            return await StreamAsChannelAsyncCore(methodName, returnType, args, cancellationToken).ForceAsync();
+            await _startLock.WaitAsync();
+            try
+            {
+                if (_connection != null)
+                {
+                    // We're already connected
+                    return;
+                }
+
+                CheckDisposed();
+
+                Log.Starting(_logger);
+
+                _connectionActive = new CancellationTokenSource();
+
+                // Check if we finished stopping (the check above proved that we were asked to stop)
+                // (We're the only ones who set _receiveTask, so this is thread-safe, because we're in our lock)
+                if (_receiveTask != null && !_receiveTask.IsCompleted)
+                {
+                    Log.WaitingForPreviousStop(_logger);
+                    // Wait for the previous stop to finish.
+                    await _receiveTask;
+                }
+
+                // Start the connection
+                var connection = _connectionFactory();
+                await connection.StartAsync(_protocol.TransferFormat);
+                _connection = connection;
+
+                Log.HubProtocol(_logger, _protocol.Name, _protocol.Version);
+
+                await HandshakeAsync();
+
+                // From here on, "ReceiveLoop" is in charge of the connection.
+                _receiveTask = ReceiveLoop();
+                Log.Started(_logger);
+            }
+            catch (Exception ex) when (_connection != null)
+            {
+                Log.ErrorStartingConnection(_logger, ex);
+                // We can't catch exceptions thrown by RecieveLoop because we don't await it, so we know it hasn't started yet.
+
+                // If we got partially started, we need to shut down the connection before we rethrow.
+                await _connection.DisposeAsync();
```

Comment:
It might be cleaner to have a smaller try catch so we know exactly how far we got when we decide this was "ok" to do. I'm a big fan of letting everything unwind then calling dispose from a single place (event loopy).

Reply:
Yeah, this needed some tidying up.